modules/photons_control/device_finder.py

```diff
@@ -541,13 +576,86 @@ def location_id(self):
             return sb.NotSpecified
         return self.location.uuid
 
+    @property
+    def product_id(self):
+        if self.product is sb.NotSpecified:
+            return sb.NotSpecified
+        return self.product.pid
+
+    @property
+    def product_name(self):
+        if self.product_id is sb.NotSpecified:
+            return sb.NotSpecified
+        return self.product.friendly
+
+    @property
+    def product_type(self):
+        if self.product is not sb.NotSpecified:
+            return (
+                ProductType.LIGHT.value
+                if Products[VendorRegistry.LIFX, self.product.pid].cap.is_light
+                else ProductType.SWITCH.value
+            )
+        else:
+            return sb.NotSpecified
+
+    @property
+    def firmware_version(self):
+        if self.firmware is sb.NotSpecified:
+            return sb.NotSpecified
+        return str(self.firmware)
+
+    @property
+    def cap(self):
+        """
+        Returns capabilities in a list as either "<cap>" or "not_<cap>".
+        and is provided for backwards compatibility. To return the original
+        capabilities as defined in products.json, use the "capabilities"
+        property
+        """
+        if self.capabilities is sb.NotSpecified:
+            return sb.NotSpecified
+
+        caps = []
+        for prop in (
+            "has_ir",
+            "has_hev",
+            "has_color",
+            "has_chain",
+            "has_relays",
+            "has_matrix",
+            "has_buttons",
+            "has_multizone",
+            "has_unhandled",
+            "has_variable_color_temp",
+        ):
+            if getattr(self.capabilities, prop):
+                caps.append(prop[4:])
+            else:
+                caps.append("not_{}".format(prop[4:]))
+        return sorted(caps)
+
+    @property
+    def capabilities(self):
+        if self.product is not sb.NotSpecified and self.firmware is not sb.NotSpecified:
+            return Products[VendorRegistry.LIFX, self.product.pid].cap(
+                self.firmware.major, self.firmware.minor
+            )
+        elif self.product is not sb.NotSpecified:
+            return Products[VendorRegistry.LIFX, self.product.pid].cap
+        else:
+            return sb.NotSpecified
+
     def as_dict(self):
         actual = super(Device, self).as_dict()
         del actual["group"]
         del actual["limit"]
         del actual["location"]
+        del actual["product"]
```

Comment:
does actual have product and firmware at this point?

Reply:
Sometimes, yes.